Fixed a few dialyzer issues and a couple of bugs found by those issues.

OvermindDL1 opened the pull request:
Overall it looks like that a lot of Phoenix's types are not accurate or are fully incorrect, it needs to be gone through in far more detail, but this is a start.  Should probably add a test on travis to fail if dialyzer is not clean or something...

ericmj commented:
Thanks for verifying with dialyzer but I don't think we want to add the `dialyxir` dependency.

OvermindDL1 commented:
@ericmj Cool, removed, just have to add it manually if that is to be used and removed then.  Or run dialyzer manually, which is comparatively painful.  ^.^

josevalim commented:
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
